src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/FlightEnabledAnalysis/FlightEnabledAbstractValue.cs

```diff
@@ -0,0 +1,176 @@
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Analyzer.Utilities.PooledObjects;
+
+#pragma warning disable CA1067 // Override Object.Equals(object) when implementing IEquatable<T> - CacheBasedEquatable handles equality
+
+namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.FlightEnabledAnalysis
+{
+    internal sealed class FlightEnabledAbstractValue : CacheBasedEquatable<FlightEnabledAbstractValue>
+    {
+        public static readonly FlightEnabledAbstractValue Unset = new FlightEnabledAbstractValue(
+            ImmutableHashSet<string>.Empty, ImmutableHashSet<FlightEnabledAbstractValue>.Empty, 0, FlightEnabledAbstractValueKind.Unset);
+        public static readonly FlightEnabledAbstractValue Empty = new FlightEnabledAbstractValue(
+            ImmutableHashSet<string>.Empty, ImmutableHashSet<FlightEnabledAbstractValue>.Empty, 0, FlightEnabledAbstractValueKind.Empty);
+        public static readonly FlightEnabledAbstractValue Unknown = new FlightEnabledAbstractValue(
+            ImmutableHashSet<string>.Empty, ImmutableHashSet<FlightEnabledAbstractValue>.Empty, 0, FlightEnabledAbstractValueKind.Unknown);
+
+        public FlightEnabledAbstractValue(
+            ImmutableHashSet<string> enabledFlights,
+            ImmutableHashSet<FlightEnabledAbstractValue> parents,
+            int height,
+            FlightEnabledAbstractValueKind kind)
+        {
+            Debug.Assert((!enabledFlights.IsEmpty || !parents.IsEmpty) == (kind == FlightEnabledAbstractValueKind.Known));
+            Debug.Assert(enabledFlights.All(enabledFlightSet => !string.IsNullOrEmpty(enabledFlightSet)));
+            Debug.Assert(parents.All(parent => parent != null));
+            Debug.Assert(height >= 0);
+            Debug.Assert(height == 0 || kind == FlightEnabledAbstractValueKind.Known);
+            Debug.Assert(height == 0 == parents.IsEmpty);
+
+            EnabledFlights = enabledFlights;
+            Parents = parents;
+            Height = height;
+            Kind = kind;
+        }
+
+        public FlightEnabledAbstractValue(string enabledFlight)
+            : this(ImmutableHashSet.Create(enabledFlight), ImmutableHashSet<FlightEnabledAbstractValue>.Empty, height: 0, FlightEnabledAbstractValueKind.Known)
+        {
+        }
+
+        public FlightEnabledAbstractValue(FlightEnabledAbstractValue parent1, FlightEnabledAbstractValue parent2)
+            : this(ImmutableHashSet<string>.Empty,
+                   ImmutableHashSet.Create(parent1, parent2),
+                   height: Math.Max(parent1.Height, parent2.Height) + 1,
+                   FlightEnabledAbstractValueKind.Known)
+        {
+        }
+
+        public ImmutableHashSet<string> EnabledFlights { get; }
```

Comment:
I plan to make the state here extensible instead of using just strings, so we can write additional analyses on top of the added DFA.